libcperciva/util/perftest.c

```diff
@@ -11,21 +11,23 @@
 
 /**
  * perftest_buffers(nbytes, sizes, nsizes, nbytes_warmup, init_func, func,
- *     cookie):
+ *     clean_func, cookie):
  * Time using ${func} to process ${nbytes} bytes in blocks of ${sizes}.
  * Before timing any block sizes, process ${nbytes_warmup} bytes with the
  * maximum size in ${sizes}.  Invoke callback functions as:
  *     init_func(cookie, buffer, buflen)
  *     func(cookie, buffer, buflen, nbuffers)
+ *     clean_func(cookie)
  * where ${buffer} is large enough to hold the maximum buffer size.  Print
- * the time and speed of processing each buffer size.
+ * the time and speed of processing each buffer size.  ${init_func} and
+ * ${clean_func} may be NULL.
  */
 int
 perftest_buffers(size_t nbytes, const size_t * sizes, size_t nsizes,
     size_t nbytes_warmup,
     int init_func(void * cookie, uint8_t * buf, size_t buflen),
     int func(void * cookie, uint8_t * buf, size_t buflen, size_t num_buffers),
-    void * cookie)
+    int clean_func(void * cookie), void * cookie)
```

Comment:
If we're going to add `clean_func` then we should probably call it in the error path, i.e. change a lot of the `goto err2` to `goto err3`?